### PR TITLE
[ESIMD] Overloading sycl sin,cos,exp,log functions for ESIMD arguments

### PR DESCRIPTION
**Pull request by DenisBakhvalov**

### CODE REVIEW

**Comment by DenisBakhvalov** [comment]
The test is here: https://github.com/intel/llvm-test-suite/pull/312

File: sycl/include/CL/sycl/builtins_esimd.hpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+#include <CL/sycl/types.hpp>
+#include <sycl/ext/intel/experimental/esimd/detail/math_intrin.hpp>
+
+// TODO Decide whether to mark functions with this attribute.
+#define __NOEXC /*noexcept*/
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+
+#define __ESIMD_NS ext::intel::experimental::esimd
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/builtins_esimd.hpp:24]: this one is already defined:
`sycl/include/sycl/ext/intel/experimental/esimd/common.hpp:#define __ESIMD_NS sycl::ext::intel::experimental::esimd`

File: sycl/include/CL/sycl/builtins_esimd.hpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ESIMD_NODEBUG ESIMD_INLINE __ESIMD_NS::simd<float, SZ>
+cos(__ESIMD_NS::simd<float, SZ> x) __NOEXC {
+#ifdef __SYCL_DEVICE_ONLY__
+  return __ESIMD_NS::detail::ocl_cos<SZ>(x.data());
+#else
+  return __esimd_cos<SZ>(x.data());
+#endif // __SYCL_DEVICE_ONLY__
+}
+
+ESIMD_NODEBUG ESIMD_INLINE __ESIMD_NS::simd<float, 1>
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply sycl/include/CL/sycl/builtins_esimd.hpp:37]: ?

**Review by kbobrovs — COMMENTED** [review]
Looks like I did not hit "submit" the first time :(

File: sycl/include/CL/sycl/builtins_esimd.hpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+#include <CL/sycl/detail/boolean.hpp>
+#include <CL/sycl/detail/builtins.hpp>
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/generic_type_traits.hpp>
+#include <CL/sycl/types.hpp>
+#include <sycl/ext/intel/experimental/esimd/detail/math_intrin.hpp>
+
+// TODO Decide whether to mark functions with this attribute.
+#define __NOEXC /*noexcept*/
```
[the remark this answers is not in the document]
Comment (DenisBakhvalov) [reply sycl/include/CL/sycl/builtins_esimd.hpp:19]: I copied it from `builtins.hpp`.

File: sycl/include/CL/sycl/builtins_esimd.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (DenisBakhvalov) [reply sycl/include/CL/sycl/builtins_esimd.hpp]: I couldn't remember why I added it. But it looks like now it's not needed. I deleted it.

File: sycl/include/CL/sycl/builtins_esimd.hpp (hunk 1 above)
Comment (DenisBakhvalov) [reply sycl/include/CL/sycl/builtins_esimd.hpp]: Done.

**Review by kbobrovs — APPROVED** [review]
please add E2E tests also

**Comment by kbobrovs** [comment]
@DenisBakhvalov, I suggest to change the title for more clarity - this is really overloading sycl::<math> functions for ESIMD arguments.

**Comment by kbobrovs** [comment]
@bader - please review. This overloading approach was agreed upon with @rolandschulz and @Pennycook.

**Comment by DenisBakhvalov** [comment]
> please add E2E tests also

E2E tests are here: https://github.com/intel/llvm-test-suite/pull/312

**Comment by kbobrovs** [comment]
@bader, @alexbatashev, @v-klochkov or other folks from rt reviewers - please approve

**Review by alexbatashev — APPROVED** [review]

**Review by dm-vodopyanov — APPROVED** [review]

**Pull request merged**